webserver/views/api.py

```diff
@@ -208,9 +96,132 @@ def get_listens(user_id):
+        raise Unauthorized("You need to provide an Authorization header.")
+    try:
+        auth_token = auth_token.split(" ")[1]
+    except IndexError:
+        raise Unauthorized("Provided Authorization header is invalid.")
+
+    user = db.user.get_by_mb_id(user_id)
+    if user is None:
+        raise Unauthorized("User %s is not known to MusicBrainz." % user_id)
+    if auth_token != user['auth_token']:
+        raise Unauthorized("Invalid authorization token.")
+
+
+def _get_listen_payload(data, raw_data):
+    try:
+        payload = data['payload']
+        if len(payload) == 0:
+            _log_raise_400("JSON document does not contain any listens.", payload)
+
+        if len(raw_data) > len(payload) * MAX_LISTEN_SIZE:
+            _log_raise_400("JSON document is too large. In aggregate, listens may not "
+                           "be larger than %d characters." % MAX_LISTEN_SIZE, payload)
+
+        if data['listen_type'] not in ('playing_now', 'single', 'import'):
+            _log_raise_400("JSON document requires a valid listen_type key.", payload)
+
+        if (data['listen_type'] == "single" or data['listen_type'] == 'playing_now') and len(payload) > 1:
+            _log_raise_400("JSON document contains more than listen for a single/playing_now. "
+                           "It should contain only one.", payload)
+    except KeyError:
+        _log_raise_400("Invalid JSON document submitted.", raw_data)
+
+    return payload
+
+
+def get_messybrainz_data(listen):
+    messy_dict = {
+        'artist': listen['track_metadata']['artist_name'],
+        'track': listen['track_metadata']['track_name'],
+    }
+    if 'release_name' in listen['track_metadata']:
+        messy_dict['release'] = listen['track_metadata']['release_name']
+
+    messy_data = json.dumps(messy_dict)
+    req = urllib2.Request(current_app.config['MESSYBRAINZ_SUBMIT_URL'], messy_data, {
+        'Content-Type': 'application/json',
+        'Content-Length': len(messy_data),
+    })
+
+    try:
+        f = urllib2.urlopen(req, timeout=current_app.config['MESSYBRAINZ_TIMEOUT'])
+        response = f.read()
+        f.close()
+    except urllib2.URLError as e:
+        current_app.logger.error("Error calling MessyBrainz:" + str(e))
+    except socket.timeout:
+        current_app.logger.error("Timeout calling MessyBrainz.")
+
+    try:
+        messy_response = json.loads(response)
+    except ValueError as e:
+        current_app.logger.error("MessyBrainz parse error: " + str(e))
+        raise InternalServerError
+
+    try:
+        messybrainz_id = messy_response['messybrainz_id']
+    except KeyError:
+        current_app.logger.error("MessyBrainz did not return a proper id")
+        raise InternalServerError
+
+    recording_id = messy_response.get('recording_id', None)
+    artist_id = messy_response.get('artist_id', None)
+    release_id = messy_response.get('release_id', None)
+
+    return messybrainz_id, recording_id, artist_id, release_id
```

Comment:
The content in here may change in the future. It may be better to just return the `messy_response` directly